### DISCUSSION TITLE
Problem using expand('%:p:h') for ripgrep in Vim

### DISCUSSION THREAD

**Discussion by stefanos82** [BurntSushi/ripgrep#2923] [Q&A]
I have the following settings in my Vim config file:

```vim

" Using fzy command for fuzzy finder "
function! FzyCommand(choice_command, vim_command)
    try
        let output = system(a:choice_command . " | fzy ")
    catch /Vim:Interrupt/
        " Swallow errors from ^C, allow redraw! below
    endtry
    redraw!
    if v:shell_error == 0 && !empty(output)
        exec a:vim_command . ' ' . output
    endif
endfunction


let findcwd="ag " . expand('%:p:h') . " --hidden --silent -l -g ''"
nnoremap <C-p> :call FzyCommand(findcwd, ":e")<cr>
```

These of course are the settings for silver-searcher.

If I replace `findcwd` configuration with `"rg " . expand('%:p:h') . " -l -g ''"` it's not working; if I replace `expand('%:p:h')` with `.` it works.

Do we know why `expand()` is not working with ripgrep?

**Comment by BurntSushi** [comment]
While I'm a (neo)vim user, I'm not an expert in vimscript, and this seems more like a vimscript question. It hinges on what `expand('%:p:h')` outputs. You also don't say what "not working" means.

My guess is that ripgrep's and ag's tty detection on stdin are slightly different (please search the issues, it is intentional) and that somehow `expand('%:p:h')` is expanding to the empty string. ripgrep then tries to search stdin where as ag searches the current working directory. By using `.` explicitly, you tell ripgrep to search the current working directory and skip any tty detection on stdin.

I also use fzf with ripgrep in neovim, but my config is a lot simpler than yours: https://github.com/BurntSushi/dotfiles/blob/01f959e5bc084ad61ec4df862d5c325041bc42ed/.config/nvim/include/keybinds.vim#L56-L60

It's unclear whether you can use the same kind of simpler config.

**Comment by stefanos82** [comment]
With the ripgrep settings configured, the mapped key `Ctrl+P` does not suggest anything apart from `>` on the bottom of editor, whereas with silver-searcher settings I get the desired result:

Silver-searcher output:

![image](https://github.com/user-attachments/assets/15734a5c-97be-4f1b-a180-a2e7419603a5)

ripgrep output:

![image](https://github.com/user-attachments/assets/5b9a0487-aae7-465a-a4fd-46e361304f8a)
